server/src/main/java/org/eclipse/openvsx/util/SnapshotHandler.java

```diff
@@ -0,0 +1,129 @@
+/********************************************************************************
+ * Copyright (c) 2023 Ericsson and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ********************************************************************************/
+package org.eclipse.openvsx.util;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.FileHandler;
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+import java.util.regex.Pattern;
+
+public class SnapshotHandler extends FileHandler {
+
+  private static final Pattern pattern = Pattern.compile("\\{.*\\}[\\r|\\n]");
+  // the following can be configured by Logging.properties
+  private final int maxEvents = 1000000;
+  private final double timeout = 30.0;
+  protected final String filePath = "request-";
+  // Enable scope logs by default
+  private volatile boolean fIsEnabled = true;
+
+  private Deque<InnerEvent> fData = new ArrayDeque<>();
+  private Map<String, Map<String, List<InnerEvent>>> fStacks = new HashMap<>();
+
+  public SnapshotHandler() throws IOException, SecurityException {
+    super();
+  }
+
+  @Override
+  public boolean isLoggable(LogRecord record) {
+    boolean logIt =
+        (fIsEnabled
+            && super.isLoggable(record)
+            && (record.getLevel().intValue() <= Level.FINE.intValue())
+            && (pattern.matcher(record.getMessage()).matches()));
+    if (logIt) { // add feature switch here
+      addToSnapshot(record.getMessage());
+    }
+    return logIt;
+  }
+
+  private boolean addToSnapshot(String message) {
+    InnerEvent event = InnerEvent.create(message);
```

Comment:
I am not sure serializing parameters to a message and then serializing them is good idea, both from the performance and the design point of view. What do you think about [LogRecord.parameters](https://docs.oracle.com/en/java/javase/21/docs//api/java.logging/java/util/logging/LogRecord.html#setParameters(java.lang.Object%5B%5D)), will it be able to do the job?

Reply:
Yes, I will update and check. The caveat: there's no "supplier" method with paramters, which means the string is always made AFAIK. That is why I made the string with the supplier. In other applications it made more sense, in VSX, not so much.

Reply:
I looked at both parameters and suppliers. For today, suppliers are quite simpler. But I am not against doing parameters if it's a strong opinion on your side. If you want to do parameters later, I am not against at all too.

Reply:
I fixed it. Now the parameters are lazy-serialized on getMessage, in the thread to drain. I think this is as good as it gets. ;)